Implement vendor-bin for owncloud-codestyle phan php_codesniffer phpstan

phil-davis opened the pull request:
## Description
- move ``owncloud/coding-standard`` ``squizlabs/php_codesniffer`` ``phan/phan`` and ``phpstan/phpstan`` into ``vendor-bin`` like has been done in apps
- adjust ``Makefile`` for that
- ``composer update``

Note: reminder to self, ``stable10`` backport will need run code style and phan in appropriate PHP combinations for PHP 5.6, 7.*

## Related Issue
Part of issue #33902 

## Motivation and Context
Be consistent with what is happening in apps.

## How Has This Been Tested?
Local runs of:
```
make test-php-style
make test-php-phan
make test-php-phpstan
```

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests

## Checklist:
- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:
- [ ] Backport (if applicable set "backport-request" label and remove when the backport was done) #33917

phil-davis committed:
Implement vendor-bin for owncloud-codestyle phan php_codesniffer phpstan

phil-davis commented on vendor-bin/phan/composer.json:
```diff
@@ -0,0 +1,5 @@
+{
+    "require": {
+        "phan/phan": "0.12.11"
```
Note: ``phan`` is pinned to ``0.12.11`` because currently any higher version reports extra stuff. That needs to be investigated as a separate issue to this PR.

phil-davis commented:
@patrickjahns @PVince81 @DeepDiver1975 this seems to work. What did I miss?

patrickjahns commented:
should we drop codesniffer for the time being?

phil-davis commented:
> should we drop codesniffer for the time being?

Our acceptance test code meets the codesniffer settings that are already there. Having this job helps us QA guys make sure that we did reasonably-correct PHP doc blocks. It does not scan any other code, so does not effect run-time code ATM.

phil-davis commented:
Also, I lost codecov -0.01% even though I did not touch any real code. If someone approves this, then if you have the privilege, please override the codecov result and merge.

phil-davis commented:
@patrickjahns I just rebased to pickup other recent composer changes in master.
Please review again.

phil-davis commented:
Dumb drone error https://drone.owncloud.com/owncloud/core/13739/256
``fatal: The remote end hung up unexpectedly``
I restarted it.

phil-davis commented:
@patrickjahns The apps have their own ``vendor-bin`` with the code style, ``phan``, ``phpstan`` ... so at the moment we are sucking in all those ``vendor-bin`` things independently via the app Makefile, composer...

patrickjahns commented on vendor-bin/owncloud-codestyle/composer.json:
```diff
@@ -0,0 +1,11 @@
+{
+    "config" : {
+      "platform": {
+        "php": "5.6.37"
```
what is the reason for pinning it to a specific php version?

phil-davis replied:
For core master, no reason. Thanks for spotting.
We do it in apps and stable10 because of wanting to make sure we get dependency versions that can still work with PHP 5.6
I pushed a change to remove this pinning.

patrickjahns commented:
Another though that crossed my mind:
we might need to update the `make clean` rules to also cleanout vendor-bin.

use-case:
The developer starts to develop on php7.1 and runs phan/phpstan once. All dependencies will be fetched that work on php7.1 (for vendor-bin) - now the developer switches to php7.3 - some phan/phpstan dependencies might need updating to work on php7.3.

phil-davis committed:
clean vendor-bin with an rm command

phil-davis commented:
@patrickjahns the 2nd commit here will clean ``vendor-bin`` when doing ``make clean`` or ``make clean-composer-deps``
Does that seem OK?
Or should I have a different make target for that?

phil-davis commented:
@patrickjahns I rebased this. Please review again.
If you are happy with the 1-line in ``Makefile`` that cleans ``vendor-bin`` then I think this is ready.

patrickjahns commented on Makefile:
```diff
@@ -147,6 +146,7 @@ update-composer: $(COMPOSER_BIN)
 clean-composer-deps:
 	rm -f $(COMPOSER_BIN)
 	rm -Rf lib/composer
+	git clean -ffdx vendor-bin
```
I am bit hesitant to use `git` here - the other clean parts just fall back to more system defaults

patrickjahns replied:
what about `rm -Rf vendor-bin/**/vendor vendor-bin/**/composer.lock` ?

phil-davis replied:
that works

phil-davis commented:
ToDo: resolve ``composer.lock``` conflicts after we get other dependency bumps merged to ```master``` (e.g. recent symfony bump) - otherwise there will just be more conflicts again.

phil-davis commented:
Rebased and re-made ``composer.lock`` just now, to pick up the composer dependencies updated in ``master`` just recently.

phil-davis commented:
Rebased and re-made ``composer.lock``, to pick up the composer dependencies updated in ``master`` just recently. I rebased on top of PR #34058 because that has other composer bumps that are waiting - makes it easier to just ```composer update``` on top of that.

phil-davis commented:
After the recent drone trouble, and other composer bumps, I have rebased again.
drone is passing.
@patrickjahns @PVince81 and/or @DeepDiver1975 please advise what I need to do next to progress this. (e.g. someone reviews, approves and merges it? ;)

phil-davis commented:
See PR #34191 for a new version of the same code. That PR passes codecov, this PR gets a silly codecov/project fail.

phil-davis commented:
Code has been merged in PR #34191
